pkg/stores/common/db.go

```diff
@@ -99,30 +154,162 @@ func InitializeUserRating(ctx context.Context, tx pgx.Tx, userId int64) error {
+	var playerRating *entity.SingleRating
+	err = tx.QueryRow(ctx, "SELECT ratings->'Data'->$1 FROM profiles WHERE user_id = $2", ratingKey, userId).Scan(&playerRating)
 	if err == pgx.ErrNoRows {
-		return nil, fmt.Errorf("profile not found for user_id: %d", userId)
+		return nil, fmt.Errorf("ratings not found for user_id: %d", userId)
 	}
 	if err != nil {
 		return nil, err
 	}
 
-	if sr == nil {
-		sr = defaultRating
-		err = UpdateUserRating(ctx, tx, userId, ratingKey, sr)
+	if playerRating == nil {
+		playerRating = entity.NewDefaultRating(true)
+		err = UpdateUserRating(ctx, tx, userId, ratingKey, playerRating)
 		if err != nil {
 			return nil, err
 		}
 	}
 
-	return sr, nil
+	return playerRating, nil
+}
+
+func GetUserStats(ctx context.Context, tx pgx.Tx, userId int64, ratingKey entity.VariantKey) (*entity.Stats, error) {
+	err := InitializeUserStats(ctx, tx, userId)
+	if err != nil {
+		return nil, err
+	}
+
+	var stats *entity.Stats
+	err = tx.QueryRow(ctx, "SELECT stats->'Data'->$1 FROM profiles WHERE user_id = $2", ratingKey, userId).Scan(&stats)
+	if err == pgx.ErrNoRows {
+		return nil, fmt.Errorf("stats not found for user_id: %d", userId)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	if stats == nil {
+		stats = &entity.Stats{}
+		err = UpdateUserStats(ctx, tx, userId, ratingKey, stats)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return stats, nil
+}
+
+func Update(ctx context.Context, tx pgx.Tx, columns []string, args []interface{}, cfg *CommonDBConfig) error {
+	for i := 0; i < len(columns); i++ {
+		columnName := columns[i]
+		columns[i] = fmt.Sprintf("%s = $%d", columnName, i+1)
+	}
+
+	setUpdatedAtStmt := ""
+	if cfg.SetUpdatedAt {
+		setUpdatedAtStmt = " updated_at = NOW(), "
+	}
+	query := fmt.Sprintf("UPDATE %s SET %s %s WHERE %s = $%d", TableTypeToString[cfg.TableType], setUpdatedAtStmt, strings.Join(columns, ","), SelectByTypeToString[cfg.SelectByType], len(columns)+1)
+	args = append(args, []interface{}{cfg.Value}...)
+	result, err := tx.Exec(ctx, query, args...)
+	if err != nil {
+		return err
+	}
+	rowsAffected := result.RowsAffected()
+	if rowsAffected != 1 {
+		return entity.NewWooglesError(ipc.WooglesError_USER_UPDATE_NOT_FOUND, fmt.Sprintf("%v", cfg.Value))
+	}
+
+	return nil
+}
+
+func GetUserBy(ctx context.Context, tx pgx.Tx, cfg *CommonDBConfig) (*entity.User, error) {
+	var id uint
+	var username string
+	var uuid string
+	var email string
+	var password string
+	internal_bot := &sql.NullBool{}
+	is_admin := &sql.NullBool{}
+	is_director := &sql.NullBool{}
+	is_mod := &sql.NullBool{}
+	var notoriety int
+	var actions *entity.Actions
+
+	placeholder := "$1"
+
+	if cfg.SelectByType == SelectByEmail || cfg.SelectByType == SelectByUsername {
+		placeholder = "lower($1)"
+	}
+
+	query := fmt.Sprintf("SELECT id, username, uuid, email, password, internal_bot, is_admin, is_director, is_mod, notoriety, actions FROM users WHERE %s = %s", SelectByTypeToString[cfg.SelectByType], placeholder)
```

Comment:
noooo! if GetUserBy gets called with user-generated input, which it does (when getting a profile, at least) this is a SQL injection. I think to be safe just don't use fmt.Sprintf with any queries; you can do `tx.QueryRow(ctx, query, values...)` where values is a slice of all the parameters, and query just contains the `$1` etc  markers.